### DISCUSSION TITLE
More advanced meal planner rules

### DISCUSSION THREAD

**Discussion by TheSuperBeaver** [mealie-recipes/mealie#4014] [Ideas]
By more advanced meal planner rules I have multiple facets in mind : 

1. Improve the current model to include food preferences for the group members
2. Have a way to manage a household calendar, this to be able to specify which household member will eat a cooked meal that day.
3. Define more advanced meal planner rules
4. Be able to generate a whole meal plan for a specific period

### 1. Food preferences : 
Be able to create links between a specific member, and an ingredient / category / tag / recipe
As an example : 
The family has 4 members
- The dad likes italian food, loves french cuisine, dislikes eggplant and artichoke
- The mom likes chinese, loves japanese cuisine, and is allergic to peanuts and nuts
- The girl is vegan, and loves green vegetables
- The boy is muslim (don't ask me why) and so can only eat beef, chicken, lamb and mutton, he also loves maroccan cuisine

For this, a new table, with corresponding services to access it needs to be created, the table would look like : 

```
FOOD_PREFERENCES : 
- ID : autogenerated UUID
- USER_ID : foreign key to USERS
- PREFERENCE : an enum that could be : LOVE, LIKE, DISLIKE, HATE, ALLERGIC, CANNOT
- A link to a tag / category / ingredient / recipe (I'm gonna need some help, there a multiple ways of doing this, my first instinct would be to have a LINK_TYPE (tag / category / ingredient / recipe) with multiple columns to the various tables with each a foreign_key constraint)
```

### 2. Household calendar
In this family of 4, the boy is there only one week every two weeks (remarriage and stuff), the dad eats outside monday and friday, and the girl has sometimes dinner over at grandma.

To be able to define this schedule, a new concept has to be added, the one of calendar, where we can define who eats which meal. By default, everyone eats every meal, and in the auto-generator, this will be taken into account to be able to calculate the needed articles on the shopping list

I visualise this as a calendar of the current month, with clickable days where you can setup exceptions, and rules you can create to specify that a person has repeating events, it could be one dinner every two weeks, or to define a shared custody.

I need to think a bit more on the actual model / table design / visuals

### 3. More advanced planner rules
This could be as simple as "activate household calendar", but here you could also define the number of times you would like to eat a specific meal per month, if you want to generate only dinners, or other settings that could be useful for the plan generation

### 4. Automatic meal plan generation
- In the meal plan section when you modify the meal plan, I envision a button to generate a meal plan for a specific period that you define in a dialog. 
- When you click on the generate button, a backend task is executed, and there, I wanted to use a tool like optapy (https://www.optapy.org/) to solve the constraints defined in the food preferences, and accounting who is there those days.
- You can also fine tune directly on the meal plan, by removing a specific meal and re-running the solver.

**Comment by camillerimichael** [comment]
I would like to add to this list if you don't mind.

- Seasonal: I would like to add a meal planner rule to only choose a soup in the colder months, while cold pasta should only be suggested in summer
- Multi select tags: A rule that can OR would be super helpful. I have tagged recipes as Mexican, Indian etc. My intention is to choose a dish on a select day from the tagged cuisines.
- Multi select days would also be helpful to reduce rule clutter. I currently have 3 days with the same rule making it harder to maintain

**Comment by ProperTeaMogul** [comment]
This is a great idea. Some ideas on implementation:

1. Create a separate meal planner calendar from which it takes the relevant rules for that date. I.e. add tag/category restrictions as all-day events in the calendar. For instance, on 21st July 2027 an all day event called "tag = vegetarian".
2. Allow the creation of restriction 'groups' which allow easier assignment of restrictions against a specific day. E.g. using above example, group "boys rules" = a collection of tag/category rules around eating beef, or chicken, or lamb or mutton.
3. Allow recurring rules within the calendar so that in the above example, the group 'boy's rules' reoccur every two weeks whilst the group "dad's rules" occur every Monday and Wednesday.

**Comment by parumpum** [comment]
I'm playing around with Google's OR-Tools CP-SAT Solver for this. Makes it very simple to specify a variety of constraints. 

Currently just utilizing a playground, but I'll work on getting a very basic implementation into the Mealie that will initially just be based on recipe Tags.

My idea for the first phase:
1. No DB changes, will utilize existing Recipe data without accounting for individual preferences/restrictions.
2. If a recipe contains a tag for a meal, it will only be allowed in that meal (Breakfast, Lunch, Dinner)
3. Ensure a recipe cannot be repeated more than X times in the week.
4. Simple UI to allow specifying Tags to include and exclude.
5. Ability to add nutrition constraints like Protein >= 100g daily.

Initially would have this set to generate for the upcoming calendar week. Starting with this may get to a releasable state that can be iterated on over time to enhance with dietary restrictions and more advanced options/constraints.

**Reply by CloCkWeRX** [reply]
Did you have much luck with this approach?

**Reply by parumpum** [reply]
Have had some work on it. Mostly backend pieces, playing with different sorts of constraints. Definitely requires a large pool of recipes with high quality data to get much out of it.

Will hopefully get back to this in the next couple weeks to keep chipping away at it.

**Comment by ProperTeaMogul** [comment]
I had a bit of success with a basic workflow in n8n and Gemini where the user sends a set of instructions and preferences for the coming days and then this, along with the available tags and categories, are parsed to Gemini to create a set of SQL queries. These are then queried against the database and the resulting lists are combined and passed back to Gemini (along with the original instructions) to suggest suitable meals for each time/day. 

Happy to share this with anyone. May at some point look to implement it directly into Mealie itself.
